File: es6/prod/src/core/di/injector.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { CONST_EXPR, CONST, stringify, isBlank } from 'angular2/src/facade/lang';
import { unimplemented, BaseException } from 'angular2/src/facade/exceptions';
const _THROW_IF_NOT_FOUND = CONST_EXPR(new Object());
export const THROW_IF_NOT_FOUND = CONST_EXPR(_THROW_IF_NOT_FOUND);
class _NullInjector {
    get(token, notFoundValue = _THROW_IF_NOT_FOUND) {
        if (notFoundValue === _THROW_IF_NOT_FOUND) {
            throw new BaseException(`No provider for ${stringify(token)}!`);
        }
        return notFoundValue;
    }
}
/**
 * The Injector interface. This class can also be used
 * to get hold of an Injector.
 */
export class Injector {
    /**
     * Retrieves an instance from the injector based on the provided token.
     * If not found:
     * - Throws {@link NoProviderError} if no `notFoundValue` that is not equal to
     * Injector.THROW_IF_NOT_FOUND is given
     * - Returns the `notFoundValue` otherwise
     *
     * ### Example ([live demo](http://plnkr.co/edit/HeXSHg?p=preview))
     *
     * ```typescript
     * var injector = ReflectiveInjector.resolveAndCreate([
     *   provide("validToken", {useValue: "Value"})
     * ]);
     * expect(injector.get("validToken")).toEqual("Value");
     * expect(() => injector.get("invalidToken")).toThrowError();
     * ```
     *
     * `Injector` returns itself when given `Injector` as a token.
     *
     * ```typescript
     * var injector = ReflectiveInjector.resolveAndCreate([]);
     * expect(injector.get(Injector)).toBe(injector);
     * ```
     */
    get(token, notFoundValue) { return unimplemented(); }
}
Injector.THROW_IF_NOT_FOUND = _THROW_IF_NOT_FOUND;
Injector.NULL = new _NullInjector();
let _EmptyInjectorFactory = class _EmptyInjectorFactory {
    create(parent = null, context = null) {
        return isBlank(parent) ? Injector.NULL : parent;
    }
};
_EmptyInjectorFactory = __decorate([
    CONST(), 
    __metadata('design:paramtypes', [])
], _EmptyInjectorFactory);
/**
 * A factory for an injector.
 */
export class InjectorFactory {
    /**
     * Binds an InjectorFactory to a fixed context
     */
    static bind(factory, context) {
        return new _BoundInjectorFactory(factory, context);
    }
}
// An InjectorFactory that will always delegate to the parent.
InjectorFactory.EMPTY = CONST_EXPR(new _EmptyInjectorFactory());
class _BoundInjectorFactory {
    constructor(_delegate, _context) {
        this._delegate = _delegate;
        this._context = _context;
    }
    create(parent = null, context = null) {
        return this._delegate.create(parent, this._context);
    }
}

File: lib/src/core/di/injector.dart
library angular2.src.core.di.injector;

import "package:angular2/src/facade/lang.dart" show stringify, isBlank;
import "package:angular2/src/facade/exceptions.dart"
    show unimplemented, BaseException;

const _THROW_IF_NOT_FOUND = const Object();
const THROW_IF_NOT_FOUND = _THROW_IF_NOT_FOUND;

class _NullInjector implements Injector {
  dynamic get(dynamic token, [dynamic notFoundValue = _THROW_IF_NOT_FOUND]) {
    if (identical(notFoundValue, _THROW_IF_NOT_FOUND)) {
      throw new BaseException('''No provider for ${ stringify ( token )}!''');
    }
    return notFoundValue;
  }
}

/**
 * The Injector interface. This class can also be used
 * to get hold of an Injector.
 */
abstract class Injector {
  static var THROW_IF_NOT_FOUND = _THROW_IF_NOT_FOUND;
  static Injector NULL = new _NullInjector();
  /**
   * Retrieves an instance from the injector based on the provided token.
   * If not found:
   * - Throws [NoProviderError] if no `notFoundValue` that is not equal to
   * Injector.THROW_IF_NOT_FOUND is given
   * - Returns the `notFoundValue` otherwise
   *
   * ### Example ([live demo](http://plnkr.co/edit/HeXSHg?p=preview))
   *
   * ```typescript
   * var injector = ReflectiveInjector.resolveAndCreate([
   *   provide("validToken", {useValue: "Value"})
   * ]);
   * expect(injector.get("validToken")).toEqual("Value");
   * expect(() => injector.get("invalidToken")).toThrowError();
   * ```
   *
   * `Injector` returns itself when given `Injector` as a token.
   *
   * ```typescript
   * var injector = ReflectiveInjector.resolveAndCreate([]);
   * expect(injector.get(Injector)).toBe(injector);
   * ```
   */
  dynamic get(dynamic token, [dynamic notFoundValue]) {
    return unimplemented();
  }
}

/**
 * An simple injector based on a Map of values.
 */
class MapInjector implements Injector {
  Injector _parent;
  Map<dynamic, dynamic> _values;
  MapInjector(this._parent, this._values) {
    if (isBlank(this._parent)) {
      this._parent = Injector.NULL;
    }
  }
  dynamic get(dynamic token, [dynamic notFoundValue = _THROW_IF_NOT_FOUND]) {
    if (identical(token, Injector)) {
      return this;
    }
    return this._values.containsKey(token)
        ? this._values[token]
        : this._parent.get(token, notFoundValue);
  }
}
